feat(engine) towards improving meaningless API range queries

filiphr opened the pull request:
This is just some idea that I had regarding the improving of the meaningless API range queries. While working on it I noticed that there are equal meaningless API call as well, like for example when there is `processInstanceId` and `processInstanceIdIn`, so I took those into consideration as well.

Let me know what you think, and whether the direction is OK.

Cheers,
Filip

buildhive commented:
[camunda BPM » camunda-bpm-platform #1560](https://buildhive.cloudbees.com/job/camunda/job/camunda-bpm-platform/1560/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[camunda BPM » camunda-bpm-platform #1580](https://buildhive.cloudbees.com/job/camunda/job/camunda-bpm-platform/1580/) UNSTABLE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[camunda BPM » camunda-bpm-platform #1582](https://buildhive.cloudbees.com/job/camunda/job/camunda-bpm-platform/1582/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

ThorbenLindhauer commented:
Hi Filip,

First of all thanks for your pull request and my apologies that I did not respond earlier.

I'd like to merge your pull request since its a useful addition and I think returning an empty list without making a query is fine in case the query conditions are logically not satisfiable by any entity. 

I think we can improve the naming of the #isValid function and the methods in CompareUtil. In my opinion, an invalid query is one that we cannot execute, i.e. one where we have to throw an exception. I would prefer a naming like `hasSatisfiableConditions` (or the inverse `hasUnsatisfiableConditions`) and something similar for the methods in CompareUtil. In addition, we added a concept of query validators meanwhile that have the purpose to detect queries that should not be processed at all (and then throw an exception), so the naming might be confusing. What do you think?

Cheers,
Thorben

filiphr commented:
Hi Thorben, 

Sorry I missed your last comment. I don't have a lot of time now, but I will have a look at this during the weekend. In general I no problem in renaming the method to something more understandable. I'll have a look at the the concept of the query validators and get back to you.

Cheers,
Filip

filiphr commented:
Hi Thorben,

As I said I will look into this for the weekend. I renamed the methods to ```hasExcludingConditions```. I think it describes the method the best. What do you think about the new name?

To add from the discussion that we had on the jira ticket [CAM-4294](https://app.camunda.com/jira/browse/CAM-4294) regarding testing the method. There is one more way that we can test it without exposing it in the API. We could use Spring ```ReflectionTestUtils```, with this we can test the method directly.

If you think that this is enough I can merge all the commits into one and provide you with one commit for the merge.

Cheers,
Filip

buildhive commented:
[camunda BPM » camunda-bpm-platform #1761](https://buildhive.cloudbees.com/job/camunda/job/camunda-bpm-platform/1761/) UNSTABLE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

ThorbenLindhauer commented:
Hi Filip,

I like `hasExcludingConditions`. Regarding `CompareUtil`, I meant something different. When reading `hasExcludingConditions` it should be easy to understand what the excluding conditions are.

For example, in the method `HistoricActivityInstanceQueryImpl#hasExcludingConditions` we have the following:

```java
protected boolean hasExcludingConditions() {
    boolean excluding = super.hasExcludingConditions();
    excluding = excluding || CompareUtil.hasExcludingOrder(startedAfter, startedBefore);
    excluding = excluding || CompareUtil.hasExcludingOrder(finishedAfter, finishedBefore);
    return excluding;
  }
```

Now I'll have to read `CompareUtil` as well to understand what an excluding order actually is. Something like

```java
protected boolean hasExcludingConditions() {
    return CompareUtil.areNotInAscendingOrder(startedAfter, startedBefore);
      || CompareUtil.areNotInAscendingOrder(finishedAfter, finishedBefore);
  }
```

would make that more obvious in my opinion.

Another note on `CompareUtil`:
I'd like to reduce the code complexity of the `hasExcludingOrder` methods. I think this mostly originates in null-checking or else we could simply compare every element with its predecessor in a for-loop. Perhaps a for-loop with an additional index that tracks the last not-null element could be used here.

Regarding tests, I'm fine with the tests you provide. I always have mixed feelings about tests that are very implementation-specific since they require maintenance when the implementation changes.

I know I'm somewhat nitpicking with my remarks here. The feature is 90% done and you say you haven't got much time at the moment. I can also make these changes myself when merging, just let me know.

Cheers,
Thorben

filiphr commented:
Hi Thorben,

Thanks a lot for your remarks, I took them into consideration and renamed the ```CompareUtil``` methods a bit better. I hope they are better now.

I also simplified the complex method into a nicer for-loop. Thanks for the tip about it.

Cheers,
Filip

buildhive commented:
[camunda BPM » camunda-bpm-platform #1766](https://buildhive.cloudbees.com/job/camunda/job/camunda-bpm-platform/1766/) UNSTABLE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

ThorbenLindhauer commented:
Hi Filip,

Once again thank you for the changes and your patience. I made some minor changes and merged the pull request into master, see https://github.com/camunda/camunda-bpm-platform/commit/60bb097286de379f89bfaa7ae3995153e085930a

Cheers,
Thorben
